### DISCUSSION TITLE
how to get raw OSError's errno raises from httpcore?

### DISCUSSION THREAD

**Discussion by x0day** [encode/httpx#2015] [Potential Issue]
`errno` can help us check DNS resolve error or connection refused error, httpx make this two errors all `ConnectError`.

```
errno.ENOEXEC => DNS resove error.
errno.ECONNREFUSED => connection refused error
```

can we remove `str(exc)` from `ConnectError` or add `errno` for `httpx.NetworkError` ? 

https://github.com/encode/httpx/blob/406700317bfd83a65d78094e4ce8e9261c284399/httpx/_transports/default.py#L74-L80

or like `aiohttp` has `strerror` and `errno` both?

https://github.com/aio-libs/aiohttp/blob/5d70b8e1457b7aee1bf8788eb8ce180dd972ff72/aiohttp/client_exceptions.py#L120
```python
class ClientConnectorError(ClientOSError):
    """Client connector error.
    Raised in :class:`aiohttp.connector.TCPConnector` if
        connection to proxy can not be established.
    """

    def __init__(self, connection_key: ConnectionKey, os_error: OSError) -> None:
        self._conn_key = connection_key
        self._os_error = os_error
        super().__init__(os_error.errno, os_error.strerror)
        self.args = (connection_key, os_error)

    @property
    def os_error(self) -> OSError:
        return self._os_error
```

**Comment by lovelydinosaur** [comment]
I'm not sure.

Question for anyone - When exceptions are chained in Python, using `raise ... from ...` how is the original exception made available on the derived exception instance?

Some questions for you @x0day...

* Are you able to provide example URL cases that'll reproduce (1) the DNS failure case (2) the connection failure case? It'd be helpful to have a super easy way to reproduce, in order to dig into how we can provide better UX around the different failure modes.
* What exact tracebacks do each of the two cases currently provide?
* What exact error message strings do each of the two cases currently provide?

> can we remove str(exc) from ConnectError or add errno for httpx.NetworkError ?

Possibly. We'd need to check if it's something that's always available at the point it's being raised.

Where are the Python docs for what possible values it can have?

**Comment by x0day** [comment]
@tomchristie 

> Where are the Python docs for what possible values it can have?
 
https://docs.python.org/3/library/errno.html

> What exact tracebacks do each of the two cases currently provide?
```text
Traceback (most recent call last):
  File "/PycharmProjects/httpx/test.py", line 3, in <module>
    resp = httpx.get("http://www.google.notexists")
  File "/Library/Caches/pypoetry/virtualenvs/httpx/lib/python3.7/site-packages/httpx/_api.py", line 201, in get
    trust_env=trust_env,
  File "/Library/Caches/pypoetry/virtualenvs/httpx/lib/python3.7/site-packages/httpx/_api.py", line 110, in request
    follow_redirects=follow_redirects,
  File "/Library/Caches/pypoetry/virtualenvs/httpx/lib/python3.7/site-packages/httpx/_client.py", line 802, in request
    return self.send(request, auth=auth, follow_redirects=follow_redirects)
  File "/Library/Caches/pypoetry/virtualenvs/httpx/lib/python3.7/site-packages/httpx/_client.py", line 893, in send
    history=[],
  File "/Library/Caches/pypoetry/virtualenvs/httpx/lib/python3.7/site-packages/httpx/_client.py", line 920, in _send_handling_auth
    history=history,
  File "/Library/Caches/pypoetry/virtualenvs/httpx/lib/python3.7/site-packages/httpx/_client.py", line 954, in _send_handling_redirects
    response = self._send_single_request(request)
  File "/Library/Caches/pypoetry/virtualenvs/httpx/lib/python3.7/site-packages/httpx/_client.py", line 990, in _send_single_request
    response = transport.handle_request(request)
  File "/Library/Caches/pypoetry/virtualenvs/httpx/lib/python3.7/site-packages/httpx/_transports/default.py", line 187, in handle_request
    resp = self._pool.handle_request(req)
  File "/usr/local/Caskroom/miniconda/base/envs/py37/lib/python3.7/contextlib.py", line 130, in __exit__
    self.gen.throw(type, value, traceback)
  File "/Library/Caches/pypoetry/virtualenvs/httpx/lib/python3.7/site-packages/httpx/_transports/default.py", line 77, in map_httpcore_exceptions
    raise mapped_exc(exc) from exc
httpx.ConnectError: [Errno 8] nodename nor servname provided, or not known
```

you can check with this snippet.

```python
import aiohttp
import asyncio
import errno


async def fetch(url):
    async with aiohttp.ClientSession() as session:
        try:
            await session.get(url)
        except aiohttp.ClientConnectorError as exc:
            if exc.os_error.errno == errno.ENOEXEC:
                print("domain resolve failed.", exc.strerror)
            elif exc.os_error.errno == errno.ECONNREFUSED:
                print("connection refused.", exc.strerror)
            elif exc.os_error.errno == errno.ECONNRESET:
                print("connection reset.", exc.strerror)
            elif exc.os_error.errno == errno.ECONNABORTED:
                print("connection abort.", exc.strerror)


async def main():
    # check for dns resolve error.
    await fetch("https://www.google.notexists")

    # check for connection refused error.
    await fetch("http://localhost:55555")
    ...

if __name__ == '__main__':
    asyncio.run(main())
```

**Comment by x0day** [comment]
> Question for anyone - When exceptions are chained in Python, using raise ... from ... how is the original exception made available on the derived exception instance?

with `httocore`, I can access errno like this:

```python
import httpcore
import errno

try:
    response = httpcore.request("GET", "http://www.google.notexists/")
except httpcore.ConnectError as exc:
    if isinstance(exc.args[0], OSError):
        err = exc.args[0]
        if err.errno == errno.ENOEXEC:
            print("dns resolve failed.")
```
not pythonic but works :)

**Reply by lovelydinosaur** [reply]
From `httpx` I think accessing `exc.__context__` would allow you to dig into the underlying socket error.

It's not *super* obvious if it's useful for us to add the error number, since it's anyways presented in the message. But it's interesting, yes. Might be nice to make that available.

Note... simplest way to check this out is using `httpx` direct from the console...

```shell
$ httpx https://www.google.notexists
ConnectError: [Errno 8] nodename nor servname provided, or not known
```

And:

```shell
$ httpx http://localhost:55555
ConnectError: [Errno 61] Connection refused
```

**Reply by x0day** [reply]
currently, check with these errors, only can check with error string match, it's not pythonic and not an universal way.

```
...
if "Connection refused" in exc.message:
    ...
...
```

maybe you can consider add `errno` and `strerror` for `httpx.ConnectError`

**Reply by x0day** [reply]
BTW, `exc.__context__` can works, thank you!
